$OpenBSD: patch-src_tun_c,v 1.1 2016/06/01 21:27:20 sthen Exp $

From db31e9def3078d899e3c43065fd86ad49deb6f83 Mon Sep 17 00:00:00 2001
From: Andrew Karpow <andy@ndyk.de>
Date: Wed, 1 Jun 2016 19:10:15 +0200
Subject: [PATCH] ocserv: fix ipv6 tun control on OpenBSD

This fixes ipv6 tunnel support on OpenBSD. OpenBSD network stack doesn't
enable the multicast flag on tun devices like FreeBSD - but this is
obligatory for ipv6.

Error message without this patch:
main: tun.c:260: tun0: Error setting IPv6: Invalid argument

Signed-off-by: Andrew Karpow <andy@ndyk.de>

--- src/tun.c.orig	Thu Apr  7 22:29:50 2016
+++ src/tun.c	Wed Jun  1 22:18:11 2016
@@ -566,6 +566,30 @@ int open_tun(main_server_st * s, struct proc_st *proc)
 		strlcpy(proc->tun_lease.name, devname(st.st_rdev, S_IFCHR), sizeof(proc->tun_lease.name));
 	}
 
+#if defined(__OpenBSD__)
+	/* enable multicast for tun interface (OpenBSD) */
+	{
+		struct tuninfo inf;
+		ret = ioctl(tunfd, TUNGIFINFO, &inf);
+		if (ret < 0) {
+			e = errno;
+			mslog(s, NULL, LOG_ERR, "%s: TUNGIFINFO: %s\n",
+					proc->tun_lease.name, strerror(e));
+			goto fail;
+		}
+
+		inf.flags |= IFF_MULTICAST;
+
+		ret = ioctl(tunfd, TUNSIFINFO, &inf);
+		if (ret < 0) {
+			e = errno;
+			mslog(s, NULL, LOG_ERR, "%s: TUNSIFINFO: %s\n",
+					proc->tun_lease.name, strerror(e));
+			goto fail;
+		}
+	}
+#endif
+
 #ifdef TUNSIFHEAD
 	{
 		int i = 1;
